cranelift/codegen/src/isa/s390x/lower.rs

```diff
@@ -0,0 +1,2807 @@
+            rd: addr.clone(),
+            rn: addr.to_reg(),
+            rm: reg.clone(),
+        });
+    }
+
+    MemArg::reg(addr.to_reg(), flags)
+}
+
+//============================================================================
+// Lowering: generating constants.
+
+fn lower_constant_u64<C: LowerCtx<I = Inst>>(ctx: &mut C, rd: Writable<Reg>, value: u64) {
+    for inst in Inst::load_constant64(rd, value) {
+        ctx.emit(inst);
+    }
+}
+
+fn lower_constant_u32<C: LowerCtx<I = Inst>>(ctx: &mut C, rd: Writable<Reg>, value: u32) {
+    for inst in Inst::load_constant32(rd, value) {
+        ctx.emit(inst);
+    }
+}
+
+fn lower_constant_f32<C: LowerCtx<I = Inst>>(ctx: &mut C, rd: Writable<Reg>, value: f32) {
+    ctx.emit(Inst::load_fp_constant32(rd, value));
+}
+
+fn lower_constant_f64<C: LowerCtx<I = Inst>>(ctx: &mut C, rd: Writable<Reg>, value: f64) {
+    ctx.emit(Inst::load_fp_constant64(rd, value));
+}
+
+//=============================================================================
+// Lowering: comparisons
+
+/// Determines whether this condcode interprets inputs as signed or
+/// unsigned.  See the documentation for the `icmp` instruction in
+/// cranelift-codegen/meta/src/shared/instructions.rs for further insights
+/// into this.
+pub fn condcode_is_signed(cc: IntCC) -> bool {
+    match cc {
+        IntCC::Equal => false,
+        IntCC::NotEqual => false,
+        IntCC::SignedGreaterThanOrEqual => true,
+        IntCC::SignedGreaterThan => true,
+        IntCC::SignedLessThanOrEqual => true,
+        IntCC::SignedLessThan => true,
+        IntCC::UnsignedGreaterThanOrEqual => false,
+        IntCC::UnsignedGreaterThan => false,
+        IntCC::UnsignedLessThanOrEqual => false,
+        IntCC::UnsignedLessThan => false,
+        IntCC::Overflow => true,
+        IntCC::NotOverflow => true,
+    }
+}
+
+fn lower_icmp_to_flags<C: LowerCtx<I = Inst>>(ctx: &mut C, insn: IRInst, is_signed: bool) {
+    let ty = ctx.input_ty(insn, 0);
+    let bits = ty_bits(ty);
+    let narrow_mode = match (bits <= 32, is_signed) {
+        (true, true) => NarrowValueMode::SignExtend32,
+        (true, false) => NarrowValueMode::ZeroExtend32,
+        (false, true) => NarrowValueMode::SignExtend64,
+        (false, false) => NarrowValueMode::ZeroExtend64,
+    };
+    let inputs = [
+        InsnInput {
+            insn: insn,
+            input: 0,
+        },
+        InsnInput {
+            insn: insn,
+            input: 1,
+        },
+    ];
+    let ty = ctx.input_ty(insn, 0);
+    let rn = put_input_in_reg(ctx, inputs[0], narrow_mode);
+    let inst = if is_signed {
+        let op = choose_32_64(ty, CmpOp::CmpS32, CmpOp::CmpS64);
+        if let Some(imm) = input_matches_simm16(ctx, inputs[1]) {
+            Inst::CmpRSImm16 { op, rn, imm }
+        } else if let Some(imm) = input_matches_simm32(ctx, inputs[1]) {
+            Inst::CmpRSImm32 { op, rn, imm }
+        } else if let Some(mem) = input_matches_mem(ctx, inputs[1]) {
```

Comment:
Good point.   It actually depends on the caller -- if `lower_icmp_to_flags` is used to lower a plain `Icmp`, then it is actually correct to sink memory loads (just like for any other instruction where we're sinking memory accesses).  However, `lower_icmp_to_flags` is also used to combine longer instruction sequences, and in those cases, sinking memory loads is only legal if all intermediate instructions only have a single user (there is currently no obvious way to check for that condition in the back-end, unfortunately).  I've now added a `may_sink_memory` flag to the routine, which is set by the caller (somewhat conservatively at the moment, this can be improved once we have support for better checks).

Reply:
Right, I agree -- the issue is precisely when `icmp`'s result is matched more than once as an operand of another instruction. The fix looks good -- thanks!